CDCgov/prime-reportstream
# Experience/feature flag env vars

Introduces the ability to set feature flags using environment variables in addition to setting them via the UI.

To set feature flags using environment variables:
* use the REACT_APP_FEATURE_FLAGS environment variable
* set the feature flags you want using the same naming convention used in the frontend ui and elsewhere in the code
* to specify multiple flags, send a comma delimited list, with no spaces, ex. `flag-1,flag-2`

Note that environment level flags cannot be deleted via the UI, but will appear there.

Some implementation notes about refactoring that went into this change:
* feature flags are now accessed by components/hooks via the `useFeatureFlags` hook
* user flags (though not env level flags) are still stored in local storage
* as a result, you no longer need to refresh after loading or removing a flag via the ui in order to see changes in the site

## Test Steps:
1. start servers on master as normal
2. visit http://localhost:3000/admin/features logging in as necessary
3. make note of any flags already in use on your account
4. restart servers on this branch
5. 2. visit http://localhost:3000/admin/features logging in as necessary
6. _VERIFY_: flags present in step 3 are still present
7. add a flag, using the ui that is not currently enabled, or remove one that is currently enabled
8. _VERIFY_: without refreshing, expected app changes resulting from feature flag change in previous step are visible
9. remove all feature flags from your account using the ui
10. restart your server with `REACT_APP_FEATURE_FLAGS='user-upload' yarn start:localdev`
11. visit http://localhost:3000/admin/features logging in as necessary
12. _VERIFY_: `user-upload` flag is present, and does not display a `remove` button in the ui. Also, note that the user upload navigation and page are available.
13. add a new flag
14. _VERIFY_:without refreshing, expected app changes resulting from feature flag change in previous step are visible
10. restart your server with `REACT_APP_FEATURE_FLAGS='user-upload' yarn start:localdev`
11. visit http://localhost:3000/admin/features logging in as necessary
12. _VERIFY_: flags set in last test are still present

## Changes
- *Include a comprehensive list of changes in this PR*
- *(For web UI changes) Include screenshots/video of changes*

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
